### ISSUE TITLE
After upgrade to 2.5.4 Foxx queues does not work anymore

### ISSUE THREAD

**Issue by razvanphp** [arangodb/arangodb#1336]
Donno exactly what more details to give, it's an app that was not changed at all, it did not work on staging when deployed today, it was working on dev server (2.5.2) and after I upgraded the dev server to 2.5.4 it stopped working there too.

I get only the log of the HTTP request that adds the job to the queue, but nothing is run from there.

Any idea?

**Comment by fceller** [comment]
What entries does db._queues contain?

**Comment by jsteemann** [comment]
Might be related to https://github.com/arangodb/arangodb/issues/1286. Currently queues seem to work from within the _system database only.

**Comment by razvanphp** [comment]
````
// [ArangoCollection: 46217620]
{"_id":"46217620","_dbName":"sorter_dev"}
````

The controller code looks like:
````javascript
var JobQueue = require('queues/ProductUpdaterQueue');
JobQueue.push('sapi-updater', id);
````

the ProductUpdaterQueue:
````javascript
var queue = Foxx.queues.create('product-updater', 1);
Foxx.queues.registerJobType('sapi-updater', function (id) {
    console.log(id);
});
module.exports = queue;
````

@jsteemann the strange thing is that it worked perfectly in 2.5.2, possibly in 2.5.3 too..

**Comment by razvanphp** [comment]
Yes, it seems to work in `_system` database.

**Comment by jsteemann** [comment]
AFAIK queues and jobs are stored in system collections named `_queues` and `_jobs`. The queue manager seems to only look into these collections inside the system database, and seems to ignore them for all other databases.

**Comment by razvanphp** [comment]
But as I said, it was working before.. strange.

Anyhow it behaves really strange, I moved the app in _system collection but I got some debug console log from the old application that was deleted.
Maybe related to this: https://github.com/arangodb/arangodb/blob/2.6/CHANGELOG#L328

**Comment by jsteemann** [comment]
I think it is. 
Queues did only work by chance before, because they were saved in database-specific collections, but were looked up in a server-wide (non-database specific) cache. So yes, it might have worked in some cases, but maybe just by chance.

**Comment by razvanphp** [comment]
Hmm, unfortunately I can reproduce this in 2.6.0a2. The queues does not seem to work at all in a custom database.

Debug info:
- In the config I have: `server.threads = 2`, `scheduler.threads = 2`, `javascript.v8-contexts = 5`
- `_queues` contains 2 items, `default` and `product-updater` both with `{"maxWorkers": 1}` inside.
- `_jobs` gets one item added each time I push to the queue, but they are never picked up nor executed:
````json
{
  "data": null,
  "maxFailures": null,
  "backOff": null,
  "onSuccess": null,
  "onFailure": null,
  "created": 1432825164148,
  "modified": 1432825164148,
  "delayUntil": 1432825164148,
  "status": "pending",
  "queue": "product-updater",
  "type": "sapi-updater",
  "failures": []
}
````

How can I check if consumers (workers) are running for this?

**Comment by pluma** [comment]
The queue was previously broken for custom databases. @jsteemann fixed this earlier this week, but the job types were still a bit fickle due to a regression in 2.5 (to summarize: Foxx became lazy-loaded in 2.5 again, but job type definitions relied on Foxx not being lazy-loaded).

I spent the past two days on completely reworking the Foxx queues. The changes are backwards-compatible but old job types continue being unreliable.

See https://github.com/arangodb/foxx-mailer-postmark and https://github.com/arangodb/foxx-segment-io for two apps I've updated to the new system. I'll update the documentation to explain how these work and how to transition from the old to the new job types later this week.

In a nutshell, the new job types are just custom scripts identified by a mount point and a script name. Because of this, they can be lazy-loaded on demand and don't have to be loaded before the worker gets to them.

There's no easy way to un-break the old job types for 2.5+ without making Foxx non-lazy again, so we will probably deprecate them in the near future. We'll try to improve communication in the Foxx team to avoid issues like this in the future.

**Comment by jsteemann** [comment]
Please also note that in order to execute jobs in Foxx queues other than the _system database, the server will need to be started with option `--server.foxx-queues-system-only false`. Otherwise it won't look into the queues of other databases than _system.

**Comment by pluma** [comment]
@jsteemann Thanks for the reminder. I need to add that to the documentation as well.

**Comment by razvanphp** [comment]
Sorry @jsteemann, but this is completely unexpected. BTW @pluma , it is in the devel documentation, I just skipped it.

Anyway, so we can understand, after looking into this for a few hours, where is the `registerJobType` function storing the job types? And with what context? Are they somehow global (that's what I can understand from using the Foxx global object)? So theoretically any job queue can use the same set of job types. This makes it over-engineered, really..

I think 2.6 is a good milestone for BC, don't wait another release cycle to fix something that is working only in some cases.

**Comment by razvanphp** [comment]
````
[....] Starting arango database server: arangod/usr/sbin/arangod: unrecognized option 'server.foxx-queues-system-only'
````

So this does not seem to be included in alpha2. I understand the concern about more CPU cycles to check all databases, but then you cannot say about this feature that "Foxx job queues are database-specific", since you just push jobs to the queue and nothing happens. At least there should be a warning message in the log saying "Hey, you pushed, thank you, but this will never be executed unless you enable this in the config".

My personal feeling is that this should not be configurable, since 98% of arangodb installs have maximum 3 databases, probably just one but renamed so that it does not use `_system`. Probably the frustration of this feature not working as expected is more important than cheap CPU cycles.

**Comment by jsteemann** [comment]
@razvanphp : the `--server.foxx-queues-system-only` option has been added shortly after alpha2 was released. If I remember correctly, alpha2 was released while we were still in the middle of fixing the queues.

**Comment by razvanphp** [comment]
In alpha3 I now get the error message from #1302, with `Unknown job type`.

@pluma can you please tell me how to define them in exports as you suggested? Or maybe some quick explanation of how to do it the new way? If 2.6 stable is nnot so close, we will need a way to make it work in 2.5. Custom scripts are new in 2.6 only?

**Comment by razvanphp** [comment]
Hey guys, any news on this one?

We will deploy next week our first Foxx app that uses the queues, but we have 2.5 in production. Do you have a launch date for 2.6 stable? If not, can you please port (a)the fix into 2.5 branch?

The use case is that we need to run this async, and as soon as one job finishes, we should somehow start a new one. Can we do this in another way?

Thank you in advance!

**Comment by pluma** [comment]
The new (impossible to break, script-based) job types are in 2.6, yes, and I sadly don't see any way of backporting them.

We could probably backport the `server.foxx-queues-system-only` setting to solve the issue with non-system DBs but there's currently no way to guarantee (2.4-style) job types are defined before the queue gets to them.

**Comment by razvanphp** [comment]
Ok, then we are going to step on thin ice on this one and use 2.6 alpha 3 (installed on staging) for the little Foxx app that we need on live.

Please give me a link to the documentation of how the new jobs work and what do I have to change, as it's a little bit tricky to understand from those two apps.

**Comment by pluma** [comment]
@razvanphp a workaround for the 2.4-style job types just landed in devel and we're planning to backport it to 2.5. If you need Foxx apps to work in non-system databases too, we could also backport the multi-db support.

For a quick explanation of the new job types, see https://docs.arangodb.com/devel/Foxx/Develop/Scripts.html

I'm also writing a cookbook recipe for creating a script and using it as a job type.

**Comment by razvanphp** [comment]
My understanding from your commit is that `foxx-queues-system-only` was removed? As I said, I find this option really confusing and useless, but yes, we have separate databases depending on team.

I would expect it to work by default on all databases, so when the job manager runs, checks all databases for jobs to run. Or do you have one job manager for each database? This would be strange..

Since we will run 2.6 for this app, we are fine like that. I already changed the jobs into scripts, I have to see how to make them running continuously, with a minimum sleep time. Also I would have two actions like start jobs and stop jobs that should work, not sure if this is possible tho.

**Comment by razvanphp** [comment]
So now I'm really confused, I read this in the documentation:
`For the low-level functionality see the chapter on the Task Management module.`

I also read this chapter and it seems closer to our needs, but the queues does not seem to support <i>recurring jobs</i>, like just rerun the same job for x number of times, at a specified interval like tasks do.

Is Foxx queues using the Tasks module? It would be great if you could tell me how to achieve what I described in me previous comment.

**Comment by pluma** [comment]
We're discontinuing support of 2.4-style job types in favour of the script-based job types that will be introduced in 2.6. Sorry about the inconvenience.
